### ISSUE TITLE
Revise Versioning Scheme

### ISSUE THREAD

**Issue by ChrisBoehmCA** [zowe/zowe-cli#77]
This would probably require us to manually bump the version before merging pull requests into master, but  would make versions easier to identify and understand by the community. This would require some work on our CICD to avoid adding the `-next<date>` part of the tag

**Comment by JTonda** [comment]
Relevant Personas:
Michelle, Ravi, Emily, Nick

**Comment by gejohnston** [comment]
# A Versioning Scheme that Clarifies Product Status

The current Zowe/Imperative versioning scheme was an attempt at semantic versioning. However, as implemented, our use of a pre-release string to create a differentiating version number complicates the specification of a Zowe dependency with the proper semantic version number string. Equally important, we have no stage between our ever-changing pre-release snapshots and our rarely-changing long-term snapshots. This document proposes changes to our versioning scheme to address such shortcomings.

The existing versioning document, [MaintainerVersioning](https://github.com/zowe/zowe-cli/blob/master/docs/MaintainerVersioning.md), will be modified with the conclusions resulting from the review of this document.

The main change to our existing versioning scheme is to create a third snapshot. Our version numbers themselves would not change. We will simply provide an early access snapshot that does not have pre-release strings in their version numbers.

Below, we layout a restructured versioning scheme. Then we will articulate how we expect that new scheme to resolve various shortcomings with the current scheme. The version numbering below is not the exact sequence, it just represents possible sequences.

|   | Supported Version | Early Access Version | Beta Version |
| - | ----------------- | -------------------- | ------------ |
| Purpose <img width=900/> | Provides a stable product that changes infrequently. Supported by Broadcom support team. <img width=400/> | Provides early access to new features. Enables customers to revise their applications to use the latest Zowe capabilities. Only supported by on-line community. <img width=400/> | Provides access to completed features that have not yet been fully validated. Enables plugin developers and end-users to validate that new behaviors solve the intended problems. Only supported by on-line community. <img width=400/> |
| Tags in current version scheme | latest | Does not exist | next |
| Tag candidates for new scheme | latest, stable, supported, longterm, or other? | next, earlyaccess, or other? | beta, validate, or other? |
| Numbering | 1.0.6, 1.0.7, 1.0.8, ... | 2.0.0, 2.1.0, 2.2.0 | 2.0.0-timestamp, 2.1.0-timestamp, 2.2.0-timestamp |
|           | 2.2.0, 2.2.1. 2.2.2. ... | 2.3.0, 2.4.0, 3.0.0, 3.1.0 | 2.3.0-timestamp, ..., 3.1.0-timestamp |
|           | 3.1.0, 3.1.1, 3.1.2, ... | 3.2.0, 3.3.0, 4.0.0, ... | 3.2.0-timestamp, ..., 4.0.0-timestamp, ... |
| Git Branch name | 1.0.0 (legacy), &nbsp; 2.2, &nbsp; 3.1, &nbsp; ... &nbsp; Only keep branches N-1.x and N.x | Same as version number. As each new branch is created, the old branch can be deleted. | master. The version number is simply changed within the master branch. |
| Release Creation | New branch created from an EarlyAccess branch containing an entire targeted feature set. Thereafter, a pipeline deploys to registry with each update. | New version branch created from master when a meaningful synch point is achieved. No updates are made until the next EarlyAccess version branch is created from master. Since we do not make updates, a pipeline may ***NOT*** be needed? | After every feature is implemented or bug is fixed, a pipeline deploys to registry. |
| Frequency of release | No more than once per year for a full new release, since we only support 2 releases and we support version N-1 for 1 year after its end-of-life announcement, which occurs when when version N is released. | At Broadcom discretion. Maybe each implemented story is released, after that story has been validated in beta? Maybe after several key features have been incorporated into and validated within a beta version? Maybe once per PI (per quarter)? | Every time a story is merged into master |
| Patching | Patches are created as needed to fix customer problems. Only 3rd digit (patch) should change for these releases. | Third digit should typically be zero. Only patch due to catastrophic error. Reported bugs should typically be made in the master/beta version, since it will be released to EarlyAccess soon enough.  | Third digit should typically be zero. Fixes are typically made within an in-progress feature version number (first two digits). We let the automatically-updated pre-release string provide a differentiated version number. The only exception might be to fix a catastrophic bug in the EarlyAccess version. |

The proposed scheme is intended to correct the following issues:

1. Currently, for plugin writers to target our next release (which uses a pre-release string), they must specify the exact versions of @brightside/core and @brightside/imperative pre-release strings. Because the pre-release string changes by date, the plugin often specifies our long-term version number or an overly vague string that match every future release. One choice results in a version mismatch warning which must be ignored. The other choice provides no real guard against pairing a plugin with an incompatible version of Zowe.

    ***Resolution***: Plugin writers can target a future Zowe release with version numbers from the EarlyAccess snapshots. Since those versions do not contain pre-release strings, plugin writers can use meaningful version-matching strings (like ~2.1.3). Such a plugin version string will not have to change with each snapshot. Further, it would not necessarily have to change when Zowe and the plugin becomes a long-term version. This should avoid the warnings and confusion that occurs, when plugin developers do not constantly update their dependencies' version strings.

2. Our current versioning scheme restricts our customers' choices to either a long-term unchanging release or an non-validated pre-release, which could change daily.

    ***Resolution***: Once an EarlyAccess snapshot exists,  we should point most users to that location for the newest software. We would only point consumers to the beta version to validate that a new feature (for which the customer was waiting) solved the intended problem, or to confirm a bug fix.

3. Our current practice of using a pre-release for all early access, obscures the line between a relatively stable early access version and a beta build that may be introducing unstable behavior. Currently, a customer who wants to try some new feature (which might be a year away from being a included in a supported version) obtains the '@next' version of Zowe. (S)he gets the most recent software merged to master. The desired feature may be relatively stable from several versions back, but the customer will get additional, non-validated features and fixes. Although our automated tests have been run, no consumer has had an opportunity to validate the very latest changes.

    ***Resolution***: The new EarlyAccess category will be the primary location from which customers obtain leading-edge product. The newest version in that category may contain more new features than the one which is desired by the customer, but all features will be validated as stable.

4. Pointing customers to a beta release involves a verbose version number like 2.3.0-next.201810182147. It is unique and correct, but it is somewhat complicated to communicate.

    ***Resolution***: Most customers will use the EarlyAccess version which will have a version number like 2.3.0. Only a customer who must validate a specific new feature or fix will have to use the more complicated version number.

The following are procedural decisions that the Darkside team must agree upon:

1. Once the major version number is bumped up, we will not update that major number further for additional breaking changes in that release, until after that major release has been published as a supported release. Additional breaking changes will just update the minor version. This does ***NOT*** adhere to the semantic versioning rules. In the EarlyAccess category, the version number will fail to communicate to customers that a new different breaking change has been introduced. For example we might have the following sequence (2.3.4, 2.3.5, 3.0.0. `3.1.0`) instead of the more correct sequence (2.3.4, 2.3.5, 3.0.0. `4.0.0`).

    While there are no technical roadblocks to fully follow the semantic versioning rules, long-standing CA Support practices expect the major number of the supported releases to occur in sequential order (1.x.x, 2.x.x, 3.x.x, 4.x.x). If each breaking change updated the major number, our supported versions for the example above would be (1.x.x, 2.x.x, 4.x.x).

    The Darkside team must decide whether to:

    * Break semantic versioning rules, or
    * Break CA Support guidelines which stipulate that major version numbers for supported products are strictly sequential.

2. Darkside must only release one supported version per year, or decide that we will take one of the following actions, which contradict CA Support guidelines:

    * Support more than 2 versions of a product, contrary to recent CA Support guidelines, or
    * Discontinue support with less than 1 year of notice, contrary to CA Support Guidelines.

3. Darkside must chose the npm tag names for our three categories of release.

4. We will not make changes to any EarlyAccess snapshot. Perhaps EarlyAccess versions are simply Git tags instead of branches? We want to avoid the additional effort to make bug fixes in the EarlyAccess snapshots. Instead, we ask EarlyAccess customers to wait the short time for the next beta version (with the bug fix) to be placed into EarlyAccess. Potentially, we already have to fix the same bug in supported versions N and N-1, and in a new beta version. The decision to not make patches to EarlyAccess snapshots will reduce our patching overhead.

5. Our beta version will continue to automatically bump our version number. But for EarlyAccess, scripts should remove the pre-release string, rebuild, and publish the version to the appropriate npm tag. If we do not fix bugs in EarlyAccess, we may not need a pipeline. We could just manually run one script to perform the tasks above.

6. Darkside must identify the desired frequency of release for EarlyAccess versions.

**Comment by gejohnston** [comment]
To track changes, I completely replace the proposal above and maintain a change log here.

# Items concluded during Nov 15 review
- Reverse the order of columns in the table to show a left-to-right time progression.
- Change the name of the `Supported` column to `Long Term Support`
- Reword the text in the  `Purpose` row to better articulate why the 3 releases exist.
- Add `Target audience` row to table.
- Identify new NPM tag names as `beta`, `latest`, `LTS-current`, `LTS-previous`.
- Identify the use of a single `earlyaccess` Git branch name.
- Identify the use of Git tags for LTS releases.
- In the `Release Creation` row, state that pipelines will automatically remove the timestamp for earlyaccess, and update the patch number for LTS.
- In the `Release Creation` row, state that we initially point the `LTS-current` NPM tag to the same version as the `earlyaccess` tag.
- In the `Frequency of release` row, remove the implication that Broadcom controls the timing of earlyaccess releases.
- In the `Frequency of release` row, identify a 6 month release cycle for LTS releases and 6 month EOL statement for the previous LTS release.
- Add issue-and-resolution number 5 to communicate how Broadcom product versioning and semantic versioning will be reconciled.

# Items concluded during Nov 28 review
- Change minimum release cycle from 6 months to 1 year, and change EOL back to 1 year.


# Items that we must still decide
- LTS versions (starting with 2.0 in January) will be similar to our CA incremental release program as both non-breaking enhancements and bug fixes can be delivered
    - Could cause conflicts with semantic versioning?
    - Could create the requirement to support more than 2 releases?
- EOL but still supported versions will only receive bug fixes (maintenance mode)
    - Does this mean only bug fixes for the one year after EOL announcement?
- Confirm approach for LTS-current and LTS-previous
    - Confirm re-naming tags for the delivery of a new LTS release.
- Can we identify a milestone that moves a release from beta to earlyaccess?
    - How do we know that a beta has been validated?
    - If we receive little customer feedback, should we move one set of features from beta to earlyaccess when the next big set of features are ready for beta?
    - Should we set a time limit after which we move from beta to earlyaccess?
- Should we update patch numbers when merging into master?
- Should we avoid making patches to the EarlyAccess snapshot?
    - Should we instead ask EarlyAccess customers to wait the short time for the next beta version (with the bug fix) to be placed into EarlyAccess? Potentially, we may have to fix the same bug in LTS-current, LTS-previous, and in a new beta version. The decision to avoid patches to EarlyAccess will reduce our patching overhead.

**Comment by ChrisBoehmCA** [comment]
I guess this issue is resolved now that we have moved to the new versioning scheme, right?

**Comment by MikeBauerCA** [comment]
Thanks @ChrisBoehmCA. Closing.
